mindmeld/models/nn_utils/classification.py

```diff
@@ -0,0 +1,490 @@
+        except KeyError:  # raised in cased on sequence classification
+            pass
+
+        # use all default params and inputted params to update self.params
+        self.params.update(params)
+
+        # init the graph and move model to device, inputs are moved to device on-the-go
+        self._init_graph()
+        self.to(self.params.device)
+        self.who_am_i()
+
+        # dumping weights during training process into a temp folder instead of keeping in
+        # memory to reduce memory usage
+        temp_folder = os.path.join(USER_CONFIG_DIR, "tmp", "pytorch_models", str(uuid.uuid4()))
+        temp_weights_save_path = os.path.join(temp_folder, "pytorch_model.bin")
+        os.makedirs(temp_folder, exist_ok=True)
+
+        # split input data into train & dev splits, and get data loaders
+        indices = np.arange(len(examples))
+        np.random.seed(SEED)
+        np.random.shuffle(indices)
+        train_examples, train_labels = zip(*[
+            (examples[i], labels[i])
+            for i in indices[:int((1.0 - self.params.dev_split_ratio) * len(indices))]
+        ])
+        dev_examples, dev_labels = zip(*[
+            (examples[i], labels[i])
+            for i in indices[int((1.0 - self.params.dev_split_ratio) * len(indices)):]
+        ])
+
+        # create an optimizer and attach all model params to it
+        num_training_steps = int(
+            len(train_examples) / self.params.batch_size / self.params.gradient_accumulation_steps *
+            self.params.number_of_epochs
+        )
+        optimizer, scheduler = self._create_optimizer_and_scheduler(num_training_steps)
+
+        # training w/ validation
+        best_dev_score, best_dev_epoch = -np.inf, -1
+        msg = f"Beginning to train for {self.params.number_of_epochs} number of epochs"
+        logger.info(msg)
+        if self.params.number_of_epochs < 1:
+            raise ValueError("Param 'number_of_epochs' must be a positive integer greater than 0")
+        _patience_counter = 0
+        verbose = (
+            logger.getEffectiveLevel() == logging.INFO or
+            logger.getEffectiveLevel() == logging.DEBUG
+        )
+        for epoch in range(1, self.params.number_of_epochs + 1):
+            # patience before terminating due to no dev score improvements
+            if _patience_counter >= self.params.patience:
+                msg = f"Set patience of {self.params.patience} epochs reached"
+                logger.info(msg)
+                break
+            # set modules to train phase, reset gradients, do forward-backward propogations
+            self.train()
+            optimizer.zero_grad()
+            train_loss, train_batches = 0.0, 0.0
+            t = tqdm(
+                range(0, len(train_examples), self.params.batch_size), disable=not verbose
+            )
+            for start_idx in t:
+                this_examples = train_examples[start_idx:start_idx + self.params.batch_size]
```

Comment:
```suggestion
                batch_examples = train_examples[start_idx:start_idx + self.params.batch_size]
```

rename this_* to batch_*?

Reply:
yup, did that and should be seen in next push!